woocommerce/google-listings-and-ads
# API PULL - Add Coupons Notification

### Changes proposed in this Pull Request:

This PR adds the logic for adding Cupon Notifications in the API PULL Project.

⚠️  It follows the same structure as Product Notifications. Hence a lot of code can be refactored reducing duplications. This will be done in further PRs.

⚠️  IF your domain is not authorized for using the sandbox, this PR wont work as a successful notification it's needed to update the statuses. So one option is to hardcode the next line so the notification is always successful.

`NotificationsService.php`
```php
public function notify( string $topic, $item_id = null ): bool {
     return true;
}
```

### Detailed test instructions:


1. With the Onboarding finished, crate a coupon with the status "DRAFT" and Channel Visibility "Dont show coupon on Google"
2. Check the Action Scheduler and see there is NO action `gla/jobs/notifications/coupons/process_item`
3. Change the status to Publish but still "Dont show coupon on Google". 
4. Check the Action Scheduler and see there is NO action `gla/jobs/notifications/coupons/process_item`
5. Change the channel visibility to  "Show coupon on Google"
6. Check the Action Scheduler and see that there is an action `gla/jobs/notifications/coupons/process_item` with `coupon.create` topic.
7. Run the action 
8. Update the coupon description
9. Check the Action Scheduler and see that there is an action `gla/jobs/notifications/coupons/process_item` with `coupon.update` topic.
10. Change the channel visibility again to  "Don't Show coupon on Google"
11.  Check the Action Scheduler and see that there is an action `gla/jobs/notifications/coupons/process_item` with `coupon.delete` topic.
12. Update the coupon description
13. Check the Action Scheduler and see there is NO action `gla/jobs/notifications/coupons/process_item`
14. Change the channel visibility to  "Show coupon on Google"
15. Check the Action Scheduler and see that there is an action `gla/jobs/notifications/coupons/process_item` with `coupon.create` topic.
16. Test other scenarios to verify that the notifications are being sent correctly. 

### Additional details:


### Changelog entry

>
